OrbitQt/ConnectionArtifacts.h

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_CONNECTION_ARTIFACTS_H_
+#define ORBIT_QT_CONNECTION_ARTIFACTS_H_
+
+#include <memory>
+
+#include "DeploymentConfigurations.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitClientData/ProcessData.h"
+#include "OrbitClientServices/ProcessManager.h"
+#include "OrbitGgp/Instance.h"
+#include "OrbitSsh/Context.h"
+#include "OrbitSsh/Credentials.h"
+#include "grpcpp/channel.h"
+#include "servicedeploymanager.h"
+
+namespace orbit_qt {
+
+// A class to store remote connection related artifacts. This is used by the ProfilingTargetDialog
+// and OrbitApp to be able to use the same connection. The
+class ConnectionArtifacts {
+ public:
+  ConnectionArtifacts(const OrbitSsh::Context* ssh_context,
+                      const ServiceDeployManager::GrpcPort& grpc_port,
+                      const DeploymentConfiguration* deployment_configuration)
+      : ssh_context_(ssh_context),
+        grpc_port_(grpc_port),
+        deployment_configuration_(deployment_configuration) {
+    CHECK(ssh_context_ != nullptr);
+    CHECK(deployment_configuration_ != nullptr);
+  }
+  void CreateServiceDeployManager(OrbitSsh::Credentials credentials) {
+    CHECK(service_deploy_manager_ == nullptr);
+    service_deploy_manager_ = std::make_unique<ServiceDeployManager>(
+        deployment_configuration_, ssh_context_, std::move(credentials), grpc_port_);
+  }
+
+  // Required Member
+  const OrbitSsh::Context* ssh_context_;
```

Comment:
Why don't you use references instead of pointers if those are required to be non-null?

Reply:
I politely disagree. :smiley:

I think we shouldn't use references as non-nullable pointers. They have never been meant for that. I'm pretty sure there is a section about that in the C++ style guide but I haven't found it so far.

I think the root cause of this problem is that in C-APIs pointers are often considered nullable, but in C++ they shouldn't be by default. If you need something nullable, `std::optional<T*>` can express that a lot more explicit. Of course you never know what contract the API designer followed, that's why wrappers like `gsl::not_null<T*>` had been invented which can explicitly state the non-nullable case.

But that's just my opinion.

Reply:
References vs. pointers are mentioned in the context of function parameters here:
https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

"Input parameters should usually be values or const references, while non-optional output and input/output parameters should usually be references (which cannot be null). Generally, use absl::optional to represent optional by-value inputs, and use a const pointer when the non-optional form would have used a reference. Use non-const pointers to represent optional outputs."

IMHO this is the same case here. This member is passed in through the constructor and should be non-null when passed there. You could argue it should still be stored as a pointer, and only the constructor should be changed to a reference, which would be the case mentioned above in the style guide - fair enough, but what is the drawback of a reference here?

Reply:
I believe the mentioned rules does not apply for constructors. The main problem I see with a reference in a constructor is that the user does not know that this reference is stored internally, so the user does not know that it's their job to keep the passed-in object alive at least as long as the constructed class exists (That's exactly what needs to happen here). A raw pointer means that the lifetime needs to be managed by the caller and they have to explicitly add a `&` when a local variable is used as an argument. That makes that contract more explicit - at least to me.

Reply:
Quick Example:
```
class CSVParser {
  explicit CSVParser(const std::string&);
  ...
};

void foo() {
  CSVParser parser{readFromFile()};

  while(const auto& line : parser) {
    // do something
  }
}
```

If the `const std::string&` is not copied, but stored as a ref this won't work. But it feels correct to me. I would assume that the string is copied in that situation.

The other case:

```
class CSVParser {
  explicit CSVParser(const std::string*);
  ...
};

void foo() {
  const auto buf = readFromFile()
  CSVParser parser{&buf};

  while(const auto& line : parser) {
    // do something
  }
}
```

Reply:
Ok, I can get behind that line of thought. There's a non-null CHECK in the constructor for these parameters which will catch misuse anyways, so you can ignore my comment on this.

Reply:
We can still discuss the member variable. :smiley:

That's what I thought we would be discussing in the first place. Even if the constructor took a pointer the member variable could be a reference.

Reply:
I'd like to have this discussion. The class `ConnectionArtifacts` changed in the meantime to 3 structs, but we can talk about `StadiaConnection` and the member `const OrbitSsh::Context* ssh_context` and `const DeploymentConfiguration* deployment_configuration`. 
They are both pointers, because I only use them for the constructor of ServiceDeployManager, which takes pointers (instead of references). 
Since this is the only use case of these variables, I would leave it as is unless there is a downside I don't see. But in general I like the idea of requiring a pointer to make sure the caller knows the lifetime hierarchy, and then saving a reference internally.